core/src/main/java/org/ehcache/events/CacheEventNotificationServiceImpl.java

```diff
@@ -174,60 +243,89 @@ public boolean hasListeners() {
+  }
+
+  @Override
+  public List<CacheConfigurationChangeListener> getConfigurationChangeListeners() {
+    List<CacheConfigurationChangeListener> configurationChangeListenerList = new ArrayList<CacheConfigurationChangeListener>();
+    configurationChangeListenerList.add(new CacheConfigurationChangeListener() {
+      @Override
+      public void cacheConfigurationChange(final CacheConfigurationChangeEvent event) {
+        if (event.getProperty().equals(CacheConfigurationProperty.ADDLISTENER)) {
+          EventListenerWrapper newListener = (EventListenerWrapper)event.getNewValue();
+          CacheEventListenerConfiguration listenerConfiguration = newListener.config;
+          registerCacheEventListener(newListener.getListener(), listenerConfiguration.orderingMode(),
+              listenerConfiguration.firingMode(), listenerConfiguration.fireOn());
+        } else if (event.getProperty().equals(CacheConfigurationProperty.REMOVELISTENER)) {
+          CacheEventListener<K, V> oldListener = (CacheEventListener)event.getOldValue();
+          deregisterCacheEventListener(oldListener);
+        }
+      }
+    });
+    return configurationChangeListenerList;
+  }
+
+  @Override
+  public void fireEvents() {
+    while (!eventThreadLocal.get().isEmpty()) {
+      CacheEventWrapper cacheEventWrapper = eventThreadLocal.get().get(0); 
+      if (!cacheEventWrapper.hasFailed()) {
+        dispatchEvent(cacheEventWrapper);
+      }
+      eventThreadLocal.get().remove(0);
+      removeMapEntry((K)cacheEventWrapper.cacheEvent.getKey());
     }
+  }
 
-    @Override
-    public int hashCode() {
-      return listener.hashCode();
+  @Override
+  public void fireSuccessfulEvents() {
+    for(CacheEventWrapper cacheEventWrapper : eventThreadLocal.get()) {
+      if((cacheEventWrapper.cacheEvent.getType() == EventType.CREATED)
+        || (cacheEventWrapper.cacheEvent.getType() == EventType.UPDATED)
+        || (cacheEventWrapper.cacheEvent.getType() == EventType.REMOVED)) {
+       cacheEventWrapper.markFailed(); 
+      }
     }
-    
-    @Override
-    public boolean equals(Object other) {
-      if (!(other instanceof EventListenerWrapper)) {
-        return false;
+    fireEvents();
+  }
+  
+  private ConcurrentLinkedQueue<CacheEventWrapper> getEventQueue (K key) {
+    long mapKeyForQueue = key.hashCode() % NUMBER_OF_EVENT_QUEUES;
+    keyBasedEventQueueMap.putIfAbsent(mapKeyForQueue, new ConcurrentLinkedQueue<CacheEventWrapper>());
+    return keyBasedEventQueueMap.get(mapKeyForQueue);
+  }
+  
+  private void removeMapEntry(K key) {
+    for (CacheEventWrapper cacheEventWrapper : getEventQueue(key)) {
+      if (cacheEventWrapper.hasFailed()) {
+        getEventQueue(key).remove(cacheEventWrapper);
       }
-      EventListenerWrapper l2 = (EventListenerWrapper)other;
-      return listener.equals(l2.listener);
+    }
+    if (getEventQueue(key).isEmpty()) {
+      removeQueue(key);
     }
   }
 
+  private void removeQueue (K key) {
```

Comment:
That is not needed now that you have a limited number of queues. Also the current impl is not thread safe.